### PR TITLE
Add a gather_traces command to collect distributed tracing from components

### PR DESCRIPTION
**Pull request by josecv**
## Description

Adds distributed tracing to the various components and a gather-traces command that will collect the traces.
Also adds an upload-traces command capable of uploading them to an OTLP collector

## Checklist

 - [x] I made sure to update `./CHANGELOG.md`.
 - [ ] I made sure to add any docs changes required for my change (including release notes).
 - [x] My change is adequately tested.
 - [ ] I updated `DEVELOPING.md` with any any special dev tricks I had to use to work on this code efficiently.
 - [ ] I updated `TELEMETRY.md` if I added, changed, or removed a metric name.
 - [ ] Once my PR is ready to have integration tests ran, I posted the PR in #telepresence-dev in the datawire-oss slack so that the "ok to test" label can be applied.

### CODE REVIEW

File: charts/telepresence/values.yaml [hunk 1]
```diff
@@ -204,6 +204,15 @@ telepresenceAPI:
   # Default: 0
   port: 0
 
+
+################################################################################
+## Tracing configuration
+################################################################################
+tracing:
+  # The port on which a grpc server will be exposed for traces
+  # To diable, set tracing to an empty object ("tracing: {}")
```
Comment (raphaelreyna) [inline charts/telepresence/values.yaml]: typo on `disable`

File: cmd/traffic/cmd/agent/client.go [hunk 2]
```diff
@@ -65,7 +66,12 @@ func TalkToManager(ctx context.Context, address string, info *rpc.AgentInfo, sta
 	ctx, cancel := context.WithCancel(ctx)
 	defer cancel()
 
-	conn, err := grpc.DialContext(ctx, address, grpc.WithTransportCredentials(insecure.NewCredentials()), grpc.WithBlock())
+	conn, err := grpc.DialContext(ctx, address,
```
Comment (raphaelreyna) [inline cmd/traffic/cmd/agent/client.go:69]: does this not need to be wrapped in with `if config.AgentConfig().TracingPort != 0 { ... }`?

File: pkg/client/rootd/service.go [hunk 3]
```diff
@@ -352,6 +358,14 @@ func run(c context.Context, loggingDir, configDir string) error {
 		return err
 	}
 
+	tracer, err := tracing.NewTraceServer(c, tracing.TraceConfig{
+		ProcessID:   4,
```
Comment (raphaelreyna) [inline pkg/client/rootd/service.go]: is this not supposed to be the OS PID?

File: pkg/tracing/tracer.go [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+func (ts *TraceServer) ServeGrpc(ctx context.Context, port uint16) error {
+	opts := []grpc.ServerOption{}
+	grpcHandler := grpc.NewServer(opts...)
+	sc := &dhttp.ServerConfig{
+		Handler: http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			if r.ProtoMajor == 2 && strings.HasPrefix(r.Header.Get("Content-Type"), "application/grpc") {
+				grpcHandler.ServeHTTP(w, r)
+			} else {
+				w.WriteHeader(404)
```
Comment (raphaelreyna) [inline pkg/tracing/tracer.go]: using `http.StatusNotFound` would be more readable here.

**Review by raphaelreyna — CHANGES_REQUESTED** [review]
lgtm; just minor nitpicks regarding readability

File: pkg/client/rootd/service.go (hunk 3 above)
Comment (josecv) [reply pkg/client/rootd/service.go]: you're totally right, I didn't think this through; there's also a whole bunch of other resource stuff we can pass through, I'm gonna refactor this bit

File: cmd/traffic/cmd/agent/client.go (hunk 2 above)
Comment (josecv) [reply cmd/traffic/cmd/agent/client.go:69]: nah it's actually alright; if the config disables tracing the headers will get added but they're totally no-op, traces will just be discarded, i'm pretty sure there's no impact

File: cmd/traffic/cmd/agent/config.go [hunk 5]
```diff
@@ -76,6 +77,18 @@ func (c *config) PodIP() string {
...
+func OtelResources(ctx context.Context, c Config) []attribute.KeyValue {
+	return []attribute.KeyValue{
+		attribute.Bool("tel2.has-mounts", c.HasMounts(ctx)),
+		attribute.String("tel2.workload-name", c.AgentConfig().WorkloadName),
+		attribute.String("tel2.workload-kind", c.AgentConfig().WorkloadKind),
+		attribute.String("tel2.manager-host", fmt.Sprintf("%s:%v", c.AgentConfig().ManagerHost, c.AgentConfig().ManagerPort)),
+		attribute.Bool("tel2.manual", c.AgentConfig().Manual),
+		attribute.String("k8s.namespace", c.AgentConfig().Namespace),
+		attribute.String("k8s.pod-ip", c.PodIP()),
+	}
```
Comment (raphaelreyna) [inline cmd/traffic/cmd/agent/config.go:89]: nice

**Review by raphaelreyna — APPROVED** [review]

**Pull request merged**